yjit/src/asm/arm64/inst/reg_pair.rs

```diff
@@ -0,0 +1,216 @@
+/// The operation to perform for this instruction.
+enum Opc {
+    /// When the registers are 32-bits wide.
+    Opc32 = 0b00,
+
+    /// When the registers are 64-bits wide.
+    Opc64 = 0b10
+}
+
+/// The kind of indexing to perform for this instruction.
+enum Index {
+    StorePostIndex = 0b010,
+    LoadPostIndex = 0b011,
+    StoreSignedOffset = 0b100,
+    LoadSignedOffset = 0b101,
+    StorePreIndex = 0b110,
+    LoadPreIndex = 0b111
+}
+
+/// A convenience function so that we can convert the number of bits of a
+/// register operand directly into an Opc variant.
+impl From<u8> for Opc {
+    fn from(num_bits: u8) -> Self {
+        match num_bits {
+            64 => Opc::Opc64,
+            32 => Opc::Opc32,
+            _ => panic!("Invalid number of bits: {}", num_bits)
+        }
+    }
+}
+
+/// The struct that represents an A64 register pair instruction that can be
+/// encoded.
+///
+/// STP/LDP
+/// +-------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+
+/// | 31 30 29 28 | 27 26 25 24 | 23 22 21 20 | 19 18 17 16 | 15 14 13 12 | 11 10 09 08 | 07 06 05 04 | 03 02 01 00 |
+/// |     0  1  0    1  0  0                                                                                        |
+/// | opc                    index..... imm7.................... rt2............. rn.............. rt1............. |
+/// +-------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+
+///
+pub struct RegisterPair {
+    /// The number of the first register to be transferred.
+    rt1: u8,
+
+    /// The number of the base register.
+    rn: u8,
+
+    /// The number of the second register to be transferred.
+    rt2: u8,
+
+    /// The signed immediate byte offset, a multiple of 8.
+    imm7: i16,
+
+    /// The kind of indexing to use for this instruction.
+    index: Index,
+
+    /// The operation to be performed (in terms of size).
+    opc: Opc
+}
+
+impl RegisterPair {
+    /// Create a register pair instruction with a given indexing mode.
+    fn new(rt1: u8, rt2: u8, rn: u8, disp: i16, index: Index, num_bits: u8) -> Self {
+        Self { rt1, rn, rt2, imm7: disp / 8, index, opc: num_bits.into() }
```

Comment:
```suggestion
        assert_ne!(rt1, rt2, "behavior is unpredictable with pairs of the same register");
        Self { rt1, rn, rt2, imm7: disp / 8, index, opc: num_bits.into() }
```

```
% as     
ldp x0, x0, [sp]
<stdin>:1:9: error: unpredictable LDP instruction, Rt2==Rt
```

Or we could do this assert in each of the instruction, though I'm not sure if there are exception to this requirement among all instructions that take a register pair.

Reply:
Oh, does the A64 spec say not to use the same register twice?

Reply:
Yeah, it's in the reference manual: https://developer.arm.com/documentation/ddi0487/ha/?lang=en which doesn't seem to be available in HTML. `K1.2.17`.

Reply:
If it's alright, I'm going to put this assertion in the `mod.rs` since that's where all of the other assertions live. Once we get to this point we're basically trusting the input to already be validated (for instance we don't validate the size of the displacement).

Reply:
Sure